File: services/api-gateway/src/utils/redis.js
const redis = require('redis');
const config = require('../config');
const logger = require('./logger');

let client = null;

async function createRedisClient() {
  if (client) {
    return client;
  }

  client = redis.createClient({
    socket: {
      host: config.redis.host,
      port: config.redis.port,
    },
    password: config.redis.password,
    database: config.redis.db,
  });

  client.on('error', (err) => {
    logger.error('Redis Client Error', err);
  });

  client.on('connect', () => {
    logger.info('Redis Client Connected');
  });

  await client.connect();
  return client;
}

async function getRedisClient() {
  if (!client) {
    await createRedisClient();
  }
  return client;
}

async function closeRedisClient() {
  if (client) {
    await client.quit();
    client = null;
    logger.info('Redis Client Disconnected');
  }
}

module.exports = {
  createRedisClient,
  getRedisClient,
  closeRedisClient,
};

File: services/identity-service/internal/database/redis.go
package database

import (
	"context"
	"fmt"

	"github.com/redis/go-redis/v9"
	"github.com/sirupsen/logrus"
)

// RedisClient Redis 客户端
type RedisClient struct {
	Client *redis.Client
}

// NewRedisClient 创建新的 Redis 客户端
func NewRedisClient(addr, password string, db int) (*RedisClient, error) {
	client := redis.NewClient(&redis.Options{
		Addr:     addr,
		Password: password,
		DB:       db,
	})

	// 测试连接
	ctx := context.Background()
	if err := client.Ping(ctx).Err(); err != nil {
		return nil, fmt.Errorf("failed to connect to redis: %w", err)
	}

	logrus.Info("Redis connected successfully")

	return &RedisClient{Client: client}, nil
}

// Close 关闭 Redis 连接
func (r *RedisClient) Close() error {
	return r.Client.Close()
}
